File: python/algorithms/quicksort.py
from typing import List, Tuple, TypeVar
from random import randint

T = TypeVar('T')

def quicksort(xs: List[T]) -> List[T]:
    if len(xs) < 2:
        return xs
    lt_pivot, pivot, gt_pivot = partition(xs)
    return quicksort(lt_pivot) + [pivot] + quicksort(gt_pivot)

def partition(xs: List[T], pivot_index=None) -> Tuple[List[T], T, List[T]]:
    if len(xs) == 0:
        raise ValueError("Cannot partition empty list")
    if pivot_index is None:
        pivot_index = randint(0, len(xs) - 1)
    pivot = xs[pivot_index]
    sorted = [pivot]
    lt_pivot = []
    gt_pivot = []
    for x in xs:
        if x is pivot:
            continue
        elif x < pivot:
            lt_pivot.append(x)
        else:
            gt_pivot.append(x)
    return (lt_pivot, pivot, gt_pivot)

File: csharp/AlgorithmWorkshop/QuickSort.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AlgorithmWorkshop
{
    public static class QuickSort
    {
        public static List<double> Sort(List<double> listToSort)
        {
            if (listToSort.Count <= 1) return listToSort;
            var partition = Partition(listToSort);
            var sorted = new List<double>();
            sorted.AddRange(Sort(partition[0]));
            sorted.AddRange(partition[1]);
            sorted.AddRange(Sort(partition[2]));

            return sorted;
        }

        public static List<List<double>> Partition(List<double> listToSort, int pivotIndex = 0)
        {
            if (!listToSort.Any()) throw new Exception("List cannot be empty.");
            if (pivotIndex == 0) pivotIndex = new Random().Next(listToSort.Count - 1);
            var pivot = listToSort[pivotIndex];
            listToSort.RemoveAt(pivotIndex);
            var ltPivot = new List<double>();
            var pivotList = new List<double> {pivot};
            var gtPivot = new List<double>();
            foreach (var item in listToSort)
            {
                if (item < pivot)
                {
                    ltPivot.Add(item);
                }
                else
                {
                    gtPivot.Add(item);
                }
            }

            return new List<List<double>> {ltPivot, pivotList, gtPivot};
        }
    }
}